### PR TITLE
feat: use recent searches (#18400)

### PR DESCRIPTION
**Pull request by lasoh**
CXCDS-8708 use recent searches

### CODE REVIEW

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+export class RecentSearchesService {
+  private recentSearchesSource = new ReplaySubject<string[]>();
+  private apiAvailability = false;
+  public get recentSearches$(): Observable<string[]> {
+    this.addRecentSearchesListener();
+    return this.recentSearchesSource.asObservable();
+  }
+  checkAvailability() {
+    return interval(150).pipe(
+      concatMap((_) => of((<any>window).Y_TRACKING)),
```
Comment (Platonn) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: [MEDIUM] using `window` might break SSR

Note : review whole PR for usage of global objects like window or document

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+import { interval, Observable, of, ReplaySubject } from 'rxjs';
+import { concatMap, endWith, take, takeWhile } from 'rxjs/operators';
+
+@Injectable()
+export class RecentSearchesService {
+  private recentSearchesSource = new ReplaySubject<string[]>();
+  private apiAvailability = false;
+  public get recentSearches$(): Observable<string[]> {
+    this.addRecentSearchesListener();
+    return this.recentSearchesSource.asObservable();
```
Comment (Platonn) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: [MEDIUM]
Prone to memory leaks. Here we're creating a subscription every time a getter `recentSearches$()` is accessed.
Did you consider alternative solutions, using `async` pipe and not doing manual subscriptions?

File: integration-libs/cds/assets/translations/en/cds-recentSearches.ts
Comment (Platonn) [inline integration-libs/cds/assets/translations/en/cds-recentSearches.ts:1]: [MEDIUM]
Please put translations resources into JSON files, and then import into `index.ts` file - similar as in other libraries.

File: integration-libs/cds/src/components/recent-searches/recent-searches.module.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+import { I18nModule, UrlModule } from '@spartacus/core';
+import { RecentSearchesComponent } from './recent-searches.component';
+import { RecentSearchesService } from './recent-searches.service';
+
+@NgModule({
+  exports: [RecentSearchesComponent],
+  declarations: [RecentSearchesComponent],
+  imports: [CommonModule, I18nModule, SearchBoxModule, UrlModule, RouterModule],
+  providers: [
+    RecentSearchesService,
```
Comment (Platonn) [inline integration-libs/cds/src/components/recent-searches/recent-searches.module.ts]: [QUESTION]
is there any reason we're providing a service in module, but not in doing
`@Injectable({providedIn: 'root'}`?

File: integration-libs/cds/src/components/recent-searches/recent-searches.module.ts [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    provideOutlet({
+      id: 'RECENT_SEARCHES',
```
Comment (Platonn) [inline integration-libs/cds/src/components/recent-searches/recent-searches.module.ts]: [MINOR]
Please note that outlet names share a global namespace. So the phrase "recent searches" might be also relevant in some other search functionalites (e.g. search users, etc.) - so let's scope it to "search box"

Also to avoid magic strings, for let's move the value to an enum exported in our public api.

e.g.
```ts
export enum SearchBoxOutlets {
   RECENT_SEARCHES: 'SearchBoxOutlets.RECENT_SEARCHES`
};
```

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 1 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: @Platonn do You have any idea what we can use in that case? If we know the Y_TRACKING (recentSeraches) is on `window`. Maybe some proposition?

File: integration-libs/cds/src/components/recent-searches/recent-searches.module.ts (hunk 4 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.module.ts]: You're right! 👍

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 2 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I know, but we use recentSeraches$ in 
`switchMap` and after that we use asyncPipe inside `recent-searches-component` template (`result$`)

File: integration-libs/cds/assets/translations/en/cds-recentSearches.ts
Comment (lasoh) [reply integration-libs/cds/assets/translations/en/cds-recentSearches.ts:1]: OK 👍 -> changed

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 1 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I changed to `windowRef from @spartacus/core` instead of `window` -> What do you think?

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.html [hunk 5]
```diff
@@ -0,0 +1,40 @@
+<!--
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.html]: You don't have to add a licence header to html files

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.html [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+  ~ SPDX-FileCopyrightText: 2024 SAP Spartacus team <spartacus-team@sap.com>
+  ~
+  ~ SPDX-License-Identifier: Apache-2.0
+  ~
+  -->
+
+<ng-container *ngIf="(result$ | async)?.length">
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.html]: To avoid multiple subscription to `result$` obs, you can use here `*ngIf="(result$ | async) as results"`
and add additional `ng-container` with `*ngIf="results?.length"`
So in line 17 you can also use `*ngFor="let recentSearch of results"` without subscribing with async pipe.

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+              (phrase) =>
+                phrase.toLowerCase().indexOf(context.search.toLowerCase()) >= 0
+            )
+            .slice(0, context.maxRecentSearches ?? MAX_RECENT_SEARCHES);
+        })
+      );
+    })
+  );
+
+  outletContext$: Observable<SearchBoxOutlet>;
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:51]: missing spacing below

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+  constructor(
+    @Optional() protected outletContext: OutletContextData<SearchBoxOutlet>,
+    protected winRef: WindowRef
+  ) {}
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:55]: missing spacing below

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+  ngOnInit() {
+    this.outletContext$ = this.outletContext.context$;
+  }
+  preventDefault(ev: UIEvent): void {
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:61]: missing spacing below

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+  search: string;
+  searchBoxActive: boolean;
+  maxRecentSearches?: number;
+}
+
+const MAX_RECENT_SEARCHES = 5;
+@Component({
+  selector: 'cx-recent-searches',
+  templateUrl: './recent-searches.component.html',
+})
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:33]: Could we use `changeDetection: ChangeDetectionStrategy.OnPush,` here?

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 2 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I'm not sure about the context, but what i understand from code, in `addRecentSearchesListener()` method, if `result` or `recentPhrases` will be `falsy`, this line `this.apiAvailability = true` will not be executed, thats mean that each call of `addRecentSearchesListener()` will create new subscription in memory. I think if you need to get value only one time from `this.checkAvailability()` it should be enough to add `take(1)`

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+    );
+  }
+
+  private addRecentSearchesListener() {
+    if (!this.apiAvailability) {
+      this.checkAvailability().subscribe((result) => {
+        if (result) {
+          const recentPhrases = (<any>(
+            this.winRef.nativeWindow
+          )).Y_TRACKING?.recentSearches?.getPhrases();
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Please add spacing below

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts [hunk 12]
```diff
@@ -107,7 +113,7 @@ export class SearchBoxComponent implements OnInit, OnDestroy {
   );
 
   ngOnInit(): void {
-    this.subscription = this.routingService
+    const routeStateSubscribtion = this.routingService
```
Comment (rmch91) [inline projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts]: typo in the variable name -> `routeStateSubscription`

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts [hunk 13]
```diff
@@ -120,6 +126,21 @@ export class SearchBoxComponent implements OnInit, OnDestroy {
           this.chosenWord = '';
         }
       });
```
Comment (rmch91) [inline projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts:128]: Please add spacing below

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts [hunk 14]
```diff
@@ -120,6 +126,21 @@ export class SearchBoxComponent implements OnInit, OnDestroy {
...
+    this.subscriptions.add(routeStateSubscribtion);
+
+    const chosenWordSubsribtion =
```
Comment (rmch91) [inline projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts]: also typo in the name as above

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts [hunk 15]
```diff
@@ -120,6 +126,21 @@ export class SearchBoxComponent implements OnInit, OnDestroy {
...
+      this.searchBoxComponentService.chosenWord.subscribe((chosenWord) => {
+        this.updateChosenWord(chosenWord);
+      });
```
Comment (rmch91) [inline projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts:135]: Please add spacing below

**Review by rmch91 — CHANGES_REQUESTED** [review]

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.html (hunk 5 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.html]: 👍 -> my ide added  -> sorry ->  changed

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 1 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: WindowRef.nativeWindow has a type of `Window | undefined`. On the server it will return `undefined`.  So you need to add `?` before `.Y_TRACKING)`. Also, if `window` will be undefined, i think it will also throw error on `takeWhile` when will try to access `recentSearches`, so probably you need also `result?.recentSearches` here.

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.html (hunk 6 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.html]: good point -> thanks

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts (hunk 7 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:51]: 👍

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts (hunk 8 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:55]: 👍

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts (hunk 9 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:61]: 👍

File: integration-libs/cds/src/components/recent-searches/recent-searches.component.ts (hunk 10 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.component.ts:33]: of course :)

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 1 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I'm not sure i also understand correctly the propose for` endWith(true)`. If after 5th attempt Y_TRACKING will still be `undefined`, it will return `true`. Thats mean that we will try to execute this code 
```
    const recentPhrases = (<any>(
            this.winRef.nativeWindow
          )).Y_TRACKING?.recentSearches?.getPhrases();
```

but we don't have to, if we already know that Y_TRACKING is `undefined`.

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+          if (recentPhrases) {
+            this.recentSearchesSource.next(recentPhrases);
+
+            (<any>window).Y_TRACKING.recentSearches?.addListener(
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Probably leftover, pointing to `window` instead of `winRef`

**Review by rmch91 — CHANGES_REQUESTED** [review]

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 11 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: It's not working with `take(1)` :(

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts (hunk 14 above)
Comment (lasoh) [reply projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts]: 👍

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts (hunk 13 above)
Comment (lasoh) [reply projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts:128]: 👍

File: projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts (hunk 15 above)
Comment (lasoh) [reply projects/storefrontlib/cms-components/navigation/search-box/search-box.component.ts:135]: 👍

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 16 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: yeah, eagle eye :) thanks

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 17]
```diff
@@ -18,11 +18,12 @@ export class RecentSearchesService {
...
-    return this.recentSearchesSource.asObservable();
+    return this.recentSearchesSource;
   }
+
   checkAvailability() {
     return interval(150).pipe(
-      concatMap((_) => of((<any>this.winRef.nativeWindow).Y_TRACKING)),
+      concatMap((_) => of((<any>this.winRef.nativeWindow)?.Y_TRACKING)),
       take(5),
       takeWhile((result: any) => !result.recentSearches),
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: One thing commented earlier, if this part `of((<any>this.winRef.nativeWindow).Y_TRACKING))` will return `undefined` i think it will also throw error on takeWhile when will try to access `recentSearches`, so probably you need also `result?.recentSearches` here

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 18]
```diff
@@ -18,11 +18,12 @@ export class RecentSearchesService {
...
       endWith(true)
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: https://github.com/SAP/spartacus/pull/18400#discussion_r1479795911

**Review by rmch91 — CHANGES_REQUESTED** [review]

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 18 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: What do you think about this implementation:

```
checkAvailability() {
   return interval(150).pipe(
    concatMap(() => of((<any>this.winRef.nativeWindow)?.Y_TRACKING)),
    map(result => !!result?.recentSearches),
    take(5),
    takeWhile(val => !val, true),
    last(),
}
```

- `takeWhile(val => !val, true)`: This operator is now used to continue taking values until a `truthy` one is found, but importantly, it includes the `inclusive` parameter set to `true`. This means it will emit the last checked value (whether true or false) when it stops taking values. This change is crucial because it ensures that if `Y_TRACKING.recentSearches` is found on the first attempt, the observable will complete correctly after emitting `true`.

- `last()`: Ensures that the last emitted value is taken. This is needed to grab the last value emitted by `takeWhile` — either the first occurrence of `true` or `false` if `Y_TRACKING.recentSearches` is not found within 5 attempts.

This way `checkAvailability()` will emit only one value either `true` or `false`.

In your implementation, if recentSearches will be undefined, if you log values in subscribe you will see 5 logs of undefined and 1 true. Two problems here,  that we are not interested in this undefined values and value `true` in this case should be rather `false`

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 11 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Yeah, now i see. `this.checkAvailability()` could emit values up to 6 times. I don't think this is good idea, cause we are not actually interested in intermediate values but only in final value (that will say is `recentSearches` available or not). 

Probably should work if you implement this one https://github.com/SAP/spartacus/pull/18400#discussion_r1481421048

Please check and let me know.

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 17 above)
Comment (lasoh) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: ok, I can secure this element (`? lastSearches ) and as for the whole checkAvailability method, it doesn't work like "infinite check", but it only tries to get data from Y_TRACKING API once in max 5 intervals, and if this API doesn't exists, you should not call it endlessly.

File: integration-libs/cds/src/components/recent-searches/recent-searches.module.ts (hunk 3 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.module.ts]: As far as i can see, it's already changed. So I'm resolving this threat

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+@Injectable({ providedIn: 'root' })
+export class RecentSearchesService {
+  private readonly recentSearchesSource = new ReplaySubject<string[]>();
+  private apiAvailability = false;
+
+  constructor(protected winRef: WindowRef) {}
+
+  public get recentSearches$(): Observable<string[]> {
+    this.addRecentSearchesListener();
+    return this.recentSearchesSource;
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I'm not sure why you removed `.asObservable();` in previous commits. Now it is possible to call `.next` on `recentSearchesSource` outside of the service, it should be rather encapsulated. And in this case u declared return type as `Observable<string[]>` but in fact this getter returns `ReplaySubject<string[]>`. Let's return proper type and transform replaySubject to observable, as it was before.

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+
+@Injectable({ providedIn: 'root' })
+export class RecentSearchesService {
+  private readonly recentSearchesSource = new ReplaySubject<string[]>();
+  private apiAvailability = false;
+
+  constructor(protected winRef: WindowRef) {}
+
+  public get recentSearches$(): Observable<string[]> {
+    this.addRecentSearchesListener();
```
Comment (rmch91) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Calling a side-effect method like `addRecentSearchesListener()` inside a getter is generally not considered a best practice for several reasons:

1. **Unexpected Side Effects:** Getters are typically used to access properties or computed values and are expected to be free of side effects. When invoking a getter, consumers do not usually expect it to modify the state or initiate processes such as setting up listeners or fetching data. This can lead to unexpected behavior and makes the code harder to understand and investigate.
2.  **Repeated Calls:** Each time the `recentSearches$` getter is accessed, `addRecentSearchesListener()` is called. And this method sets up event listeners, add subscriptions, calling it multiple times can lead to multiple listeners or subscriptions being created, potentially causing memory leaks or other unintended consequences.

IMO, this method should be called outside of getter, even if it will be a public method and called from the component.

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 1 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Fixed

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 18 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: Fixed

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts (hunk 2 above)
Comment (rmch91) [reply integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: We still have unsubscribed `Subscription`, it's always a red flag for me, you are calling this method `this.addRecentSearchesListener()` by accessing `results` with `async` pipe in component, so it will call this method each time when angular will trigger the change detection for this component. As far as i see, lifecycles of the component and this service are not the same. If component will be destroyed, service will still exist, and after new component creation, new call to service could create new subscriptions.

Yes, you have a guard `this.apiAvailability`, but as i described above, if this.checkAvailability() will return falsy value, this flag will be not set to true, that's mean the new call to this method will create new Subscription.

What you could possible do here:

1. use take(1)
2. Create a guard, meaning assign this `subscription` to class member, and before each call of `this.checkAvailability().subscribe()` check if subscription exist, if yes -> unsubscribe before creating the new one.
3. Convert this method, and not call `subscribe()` inside of it, but return observable instead of it. In this case you will able to use chaining in the component and in first `switchMap` you can call `addRecentSearchesListener` and receive for example true if all is good and listeners were created successfully, and in second switchMap use `this.recentSearchesService.recentSearches$`. Thanks to this, you will know that api available, and listeners created, before access too recentSearches$ and all subscriptions will be covered by `async` pipe

Right now you are accessing `recentSearches$` even before listeners created.

```
public get recentSearches$(): Observable<string[]> {
    this.addRecentSearchesListener(); // Could take same time because of intervals
    return this.recentSearchesSource; // Will return values instantly
}
```

TL;DR:

Please add a mechanism that will save us from even the slightest possibility of creating memory leaks.

**Review by rmch91 — CHANGES_REQUESTED** [review]

**Review by rmch91 — DISMISSED** [review]

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+@Injectable({ providedIn: 'root' })
+export class RecentSearchesService {
+  private readonly recentSearchesSource = new ReplaySubject<string[]>();
+  public recentSearches$ = this.recentSearchesSource.asObservable();
+  constructor(protected winRef: WindowRef) {
+    this.addRecentSearchesListener();
+  }
+
+  checkAvailability() {
+    return interval(250).pipe(
```
Comment (kpawelczak) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts:23]: In SSR this interval here will slow down the rendering of the page by 2.5s -> 250ms * 100 (take(100)). This delay is likely to cause the SSR engine to timeout and fail to render the page altogether. I would suggest to wrap the whole function `addRecentSearchesListener` in `this.winRef.isBrowser()`.

```typescript
if (this.winRef.isBrowser()) {
this.addRecentSearchesListener();
}
```

File: integration-libs/cds/src/components/recent-searches/recent-searches.service.ts [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+
+@Injectable({ providedIn: 'root' })
+export class RecentSearchesService {
+  private readonly recentSearchesSource = new ReplaySubject<string[]>();
+  public recentSearches$ = this.recentSearchesSource.asObservable();
+  constructor(protected winRef: WindowRef) {
+    this.addRecentSearchesListener();
+  }
+
+  checkAvailability() {
```
Comment (kpawelczak) [inline integration-libs/cds/src/components/recent-searches/recent-searches.service.ts]: I'm wondering whether this function should be public. Is it utilized anywhere beyond this service?

**Review by kpawelczak — CHANGES_REQUESTED** [review]
